File: src/main/java/ml/duncte123/skybot/commands/music/RepeatCommand.java
package ml.duncte123.skybot.commands.music;

import ml.duncte123.skybot.audio.GuildMusicManager;
import ml.duncte123.skybot.audio.TrackScheduler;
import ml.duncte123.skybot.objects.command.Command;
import ml.duncte123.skybot.utils.AirUtils;
import ml.duncte123.skybot.utils.AudioUtils;
import ml.duncte123.skybot.utils.Config;
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;

public class RepeatCommand extends Command {

    /**
     * This is a check to see if the command is save to execute
     * @param args The command agruments
     * @param event a instance of {@link net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent GuildMessageReceivedEvent}
     * @return true if we are the command is safe to run
     */
    @Override
    public boolean called(String[] args, GuildMessageReceivedEvent event) {

        if(!event.getGuild().getAudioManager().isConnected()){
            event.getChannel().sendMessage(AirUtils.embedField(AirUtils.au.embedTitle, "I'm not in a voice channel, use `"+Config.prefix+"join` to make me join a channel")).queue();
            return false;
        }

        if(!event.getGuild().getAudioManager().getConnectedChannel().getMembers().contains(event.getMember())){
            event.getChannel().sendMessage(AirUtils.embedField(AirUtils.au.embedTitle, "I'm sorry, but you have to be in the same channel as me to use any music related commands")).queue();
            return false;
        }

        return true;
    }

    /**
     * This is the action of the command, the thing you want the command to to needs to be in here
     * @param args The command agruments
     * @param event a instance of {@link net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent GuildMessageReceivedEvent}
     */
    @Override
    public void action(String[] args, GuildMessageReceivedEvent event) {
        AudioUtils au = AirUtils.au;

        Guild guild = event.getGuild();
        GuildMusicManager mng = au.getMusicManager(guild);
        TrackScheduler scheduler = mng.scheduler;

        scheduler.setRepeating(!scheduler.isRepeating());

        event.getChannel().sendMessage(AirUtils.embedField(au.embedTitle, "Player was set to: **" + (scheduler.isRepeating() ? "repeat" : "not repeat") + "**")).queue();

    }

    /**
     * The usage instructions of the command
     * @return a String
     */
    @Override
    public String help() {
        // TODO Auto-generated method stub
        return "Makes the player repeat the currently playing song";
    }

    @Override
    public String getName() {
        return "repeat";
    }

}

File: src/main/kotlin/ml/duncte123/skybot/commands/music/RepeatCommand.kt
/*
 * Skybot, a multipurpose discord bot
 *      Copyright (C) 2017  Duncan "duncte123" Sterken & Ramid "ramidzkh" Khan & Maurice R S "Sanduhr32"
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package ml.duncte123.skybot.commands.music

import me.duncte123.botcommons.messaging.MessageUtils.sendMsg
import ml.duncte123.skybot.extensions.toEmoji
import ml.duncte123.skybot.objects.command.CommandContext
import ml.duncte123.skybot.objects.command.MusicCommand

class RepeatCommand : MusicCommand() {

    init {
        this.name = "repeat"
        this.aliases = arrayOf("loop")
        this.help = "Makes the player repeat the currently playing song or the entire queue"
        this.usage = "[playlist/queue/status]"
    }

    override fun run(ctx: CommandContext) {
        val mng = ctx.audioUtils.getMusicManager(ctx.guild)
        val scheduler = mng.scheduler

        if (ctx.args.size == 1) {
            when (ctx.args[0]) {
                "playlist", "queue" -> {
                    val wasLooping = scheduler.isLooping

                    scheduler.isLoopingQueue = !scheduler.isLoopingQueue
                    scheduler.isLooping = false

                    var extra = ""

                    if (wasLooping) {
                        extra = "\n(Normal looping has been turned off)"
                    }

                    sendMsg(
                        ctx,
                        "Player is now set to: **${if (scheduler.isLoopingQueue) "" else "not "}repeating the current queue**$extra"
                    )
                }

                "status" -> {
                    sendMsg(
                        ctx,
                        """Current repeat status:
                        |Repeating: **${scheduler.isLooping.toEmoji()}**
                        |Repeating queue: **${scheduler.isLoopingQueue.toEmoji()}**
                        """.trimMargin()
                    )
                }

                else -> {
                    this.sendUsageInstructions(ctx)
                }
            }

            return
        }

        val wasLoopingQueue = scheduler.isLoopingQueue

        // turn off all looping
        if (scheduler.isLoopingQueue || scheduler.isLooping) {
            scheduler.isLoopingQueue = false
            scheduler.isLooping = false
        }

        scheduler.isLooping = !scheduler.isLooping

        sendMsg(
            ctx,
            "Player is now set to: **${if (scheduler.isLooping) "" else "not "}repeating" +
                "${if (wasLoopingQueue) " the current queue" else ""}**"
        )
    }
}
